sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statsEstimation/JoinEstimation.scala

```diff
@@ -0,0 +1,316 @@
+          leftRows.max(innerRows)
+        case RightOuter =>
+          // All rows from right side should be in the result.
+          rightRows.max(innerRows)
+        case FullOuter =>
+          // T(A FOJ B) = T(A LOJ B) + T(A ROJ B) - T(A IJ B)
+          leftRows.max(innerRows) + rightRows.max(innerRows) - innerRows
+        case _ =>
+          // Don't change for inner or cross join
+          innerRows
+      }
+
+      // 3. Update statistics based on the output of join
+      val intersectedStats = if (selectivity == 0) {
+        AttributeMap[ColumnStat](Nil)
+      } else {
+        updateIntersectedStats(joinKeyPairs, leftStats, rightStats)
+      }
+      val inputAttrStats = AttributeMap(
+        leftStats.attributeStats.toSeq ++ rightStats.attributeStats.toSeq)
+      val attributesWithStat = join.output.filter(a => inputAttrStats.contains(a))
+      val (fromLeft, fromRight) = attributesWithStat.partition(join.left.outputSet.contains(_))
+      val outputStats: Map[Attribute, ColumnStat] = join.joinType match {
+        case LeftOuter =>
+          // Don't update column stats for attributes from left side.
+          fromLeft.map(a => (a, inputAttrStats(a))).toMap ++
+            updateAttrStats(outputRows, fromRight, inputAttrStats, intersectedStats)
+        case RightOuter =>
+          // Don't update column stats for attributes from right side.
+          updateAttrStats(outputRows, fromLeft, inputAttrStats, intersectedStats) ++
+            fromRight.map(a => (a, inputAttrStats(a))).toMap
+        case FullOuter =>
+          // Don't update column stats for attributes from both sides.
+          attributesWithStat.map(a => (a, inputAttrStats(a))).toMap
+        case _ =>
+          // Update column stats from both sides for inner or cross join.
+          updateAttrStats(outputRows, attributesWithStat, inputAttrStats, intersectedStats)
+      }
+      val outputAttrStats = AttributeMap(outputStats.toSeq)
+
+      Some(Statistics(
+        sizeInBytes = getOutputSize(join.output, outputRows, outputAttrStats),
+        rowCount = Some(outputRows),
+        attributeStats = outputAttrStats,
+        isBroadcastable = false))
+
+    case _ =>
+      // When there is no equi-join condition, we do estimation like cartesian product.
+      val inputAttrStats = AttributeMap(
+        leftStats.attributeStats.toSeq ++ rightStats.attributeStats.toSeq)
+      // Propagate the original column stats
+      val outputAttrStats = getOutputMap(inputAttrStats, join.output)
+      val outputRows = leftStats.rowCount.get * rightStats.rowCount.get
+      Some(Statistics(
+        sizeInBytes = getOutputSize(join.output, outputRows, outputAttrStats),
+        rowCount = Some(outputRows),
+        attributeStats = outputAttrStats,
+        isBroadcastable = false))
+  }
+
+  // scalastyle:off
+  /**
+   * The number of rows of A inner join B on A.k1 = B.k1 is estimated by this basic formula:
```

Comment:
look at the code path: `JoinEstimation.estimate` -> `InnerOuterEstimation.doEstimate` -> `joinSelectivity`, outer join will hit this branch.